### PR TITLE
std.testing: add methods tmpDirPath, getTestArgs, buildExe

### PR DESCRIPTION
**Pull request by matu3ba**
continuation of #11093 to simplify testing IPC

* use cases
  - get path to temporary directory
  - get the test arguments inside test block for reusage
  - build executables from text within test blocks, ie to test IPC
* missing conventions
  - how to name and debug test cases
  - where do simple+repititve build commands for testing belong

### CODE REVIEW

**Comment by matu3ba** [comment]
Question for next PR: Should there be a method to create a zig program from a string inside tests or where should it reside?
I woudl envision such a method to contain something along the following with an interface `buildExeInTmp(allocator, progstr, binfile) ![]const u8` and the exe as return value.
```zig
    var it = try std.process.argsWithAllocator(allocator);
    const testargs = try testing.getTestArgs(&it);
    defer it.deinit(); // no-op unless WASI or Windows

    var tmp = testing.tmpDir(.{ .no_follow = true }); // ie zig-cache/tmp/8DLgoSEqz593PAEE
    defer tmp.cleanup();
    const tmpdirpath = try testing.tmpDirPath(allocator, &tmp);
    defer allocator.free(tmpdirpath);
    const child_name = "child"; // no need for suffixes (.exe, .wasm) due to '-femit-bin'
    const suffix_zig = ".zig";
    const child_path = try fs.path.join(allocator, &[_][]const u8{ tmpdirpath, child_name });
    defer allocator.free(child_path);
    const child_zig = try mem.concat(allocator, u8, &[_][]const u8{ child_path, suffix_zig });
    defer allocator.free(child_zig);

    try tmp.dir.writeFile("child.zig", childstr);
    try testing.buildExe(testargs.zigexec, child_zig, child_path);
```
That way, building (multiple) executables to a tmp folder would be exactly 1 properly checked function from within test blocks.

Alternatively, I will implement a shortcut to obtain `child_zig` and `child_path` (that means absolute path to zig file and absolute path to binary).

**Comment by matu3ba** [comment]
Some functionality is redundant to `src/test.zig`, but unfortunately this is no available for user programs, because it is not part of `lib/std`.
Any advice how to refactor `src/test.zig` or how to proceed would be nice.

File: lib/std/testing.zig [hunk 1]
```diff
@@ -400,6 +400,56 @@ pub fn tmpDir(opts: std.fs.Dir.OpenDirOptions) TmpDir {
     };
 }
 
+/// caller owns memory
+pub fn tmpDirPath(alloc: std.mem.Allocator, tmpdir: *TmpDir) ![]u8 {
```
Comment (squeek502) [inline lib/std/testing.zig:404]: Any reason why this isn't in the `TmpDir` struct as something like:

```zig
pub fn getFullPath(self: TmpDir, alloc: std.mem.Allocator) ![]u8
```

Comment (matu3ba) [reply lib/std/testing.zig]: fixed.

**Comment by iddev5** [comment]
> Question for next PR:

Its a good suggestion but eventually similar functions would be needed for all other zig operations like run, buildLib etc.
What could instead be done is a function ``fn writeZigFileInTmp(allocator, progstr, binfile) ![]const u8`` which would return the final path of file (which is child_zig) in your test, on which we can then manually do ``std.fs.path.dirname(child_zig).?`` to get child_path.

**Review by iddev5 — APPROVED** [review]

**Comment by matu3ba** [comment]
@iddev5 With your proposed solution I get
```zig
    const zigfile_path = try tmp.writeZigFile(std.testing.allocator, tmp, childstr, child_name);
    defer std.testing.allocator.free(zigfile_path);
    const binary = zigfile_path[0 .. zigfile_path.len - 4]; // '.zig' is 4 characters
    try testing.buildExe(testargs.zigexec, zigfile_path, binary);
```
because we still need the path to the resulting binary for `buildExe`.
I think requiring to copy paste from users with an example for the subslice should be fine, since
we cant do it more efficiently than the subslice.

**Pull request merged**